library/common/extensions/filters/http/platform_extension/bridging_filter.cc

```diff
@@ -0,0 +1,115 @@
+#include "library/common/extensions/filters/http/platform_extension/bridging_filter.h"
+
+//#include <iostream>
+
+#include "envoy/server/filter_config.h"
+
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+
+#include "library/common/api/external.h"
+#include "library/common/buffer/utility.h"
+#include "library/common/http/header_utility.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace PlatformExtension {
+
+Http::FilterHeadersStatus mapStatus(envoy_filter_headers_status_t status) {
+  switch (status) {
+  case ENVOY_FILTER_HEADERS_STATUS_CONTINUE:
+    return Http::FilterHeadersStatus::Continue;
+  case ENVOY_FILTER_HEADERS_STATUS_STOP_ITERATION:
+    return Http::FilterHeadersStatus::StopIteration;
+  case ENVOY_FILTER_HEADERS_STATUS_CONTINUE_AND_END_STREAM:
+    return Http::FilterHeadersStatus::ContinueAndEndStream;
+  case ENVOY_FILTER_HEADERS_STATUS_STOP_ALL_ITERATION_AND_BUFFER:
+    return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+  default:
+    ASSERT(false, "unrecognized filter status from platform: {}");
+    return Http::FilterHeadersStatus::Continue;
+  }
+}
+
+BridgingFilterConfig::BridgingFilterConfig(
+    const envoymobile::extensions::filters::http::platform_extension::Bridging& proto_config)
+    : name_(proto_config.name()) {}
+
+BridgingFilter::BridgingFilter(BridgingFilterConfigSharedPtr config)
+    : platform_filter_(
+          static_cast<envoy_http_filter*>(Api::External::retrieveApi(config->name()))) {}
+
+Http::FilterHeadersStatus BridgingFilter::decodeHeaders(Http::RequestHeaderMap& headers,
+                                                        bool end_stream) {
+  envoy_headers in_headers = Http::Utility::toBridgeHeaders(headers);
+  envoy_filter_headers_status result = platform_filter_->on_request_headers(in_headers, end_stream, platform_filter_->context);
+  Http::FilterHeadersStatus status = mapStatus(result.status);
+  // Current platform implementations expose immutable headers, thus any modification necessitates a full copy.
+  // If the returned pointer is identical, we assume no modification was made and elide the copy here.
+  if (in_headers.headers != result.headers.headers) {
+    headers.clear();
+    for (envoy_header_size_t i = 0; i < result.headers.length; i++) {
```

Comment:
TODO for optimization

Reply:
This already elides the copy in the unmodified case.

@rebello95 and @buildbreaker felt somewhat strongly about using immutable objects in the filter callbacks, so a full copy is more or less necessary if any modification was made (and in fact, has already happened). (We could filter for a diff in linear time, but then the copy is linear, too). If we created a _custom_ builder in the callback that accumulated a diff and passed back the builder _instead_ of a realized collection, we could operate on the diff directly, but I think that's introducing quite a lot of complexity at that point. After some more reflection, I basically think there are two reasonable approaches:

1. Accept that using immutable objects implies copies will be made.
2. Switch to mutable objects like Envoy does in its own filters, opening the door for fully-bridged collections.

1) seems good enough to me for now, though I'd somewhat prefer 2) long-term. Either way I think further discussion would need to happen were we to change course.

Reply:
Yeah, I feel 1 is an acceptable approach right now. We know of this could be optimized (we also know how we want to optimize it) and we can tackle this problem later on.

Reply:
Right, so I think we should leave some trail of that long term thinking. Perhaps actually an issue is better.

Reply:
To clarify, I don't think there is really room for further optimization if we stick with immutable types. (At least anything that won't be terribly convoluted.)

So, I was trying to say is that there's maybe only a TODO for further optimization if we expect to switch to mutable types at some point in the future.

But I agree with your point about capturing context, and will create an issue and link it here.

Reply:
Done.